File: MyLearning/DsaPractice/Arrays/2D/Convert2DTo1D.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace DsaPractice.Arrays
{
    internal class Convert2DTo1D
    {
        public void Convert2DArrayTo1D()
        {
            int[,] matrix = new int[,]
            {
                { 1, 1, 1, 1 },
                { 1, 0, 1, 1 },
                { 1, 1, 0, 1 },
                { 1, 0, 0, 1 }
            };

            int rows = matrix.GetLength(0);
            int cols = matrix.GetLength(1);

            int[] result = new int[matrix.Length];
            int index = 0;

            for (int row = 0; row < rows; row++)
            {
                for (int col = 0; col < cols; col++)
                {
                    result[index] = matrix[row, col];
                    index++;
                }
            }

            Console.WriteLine(string.Join(", ", result));
        }
        public void Convert1DArrayTo2D()
        {
            int[] flatArray = new int[]
            {
                1, 1, 1, 1,
                1, 0, 1, 1,
                1, 1, 0, 1,
                1, 0, 0, 1
            };

            int rows = 4;
            int cols = 4;
            int index = 0;

            int[,] result = new int[rows, cols];
            
            for (int row = 0; row < rows; row++)
            {
                for (int col = 0; col < cols; col++)
                {
                    result[row, col] = flatArray[index];
                    index++;
                }
            }

            for (int row = 0; row < rows; row++)
            {
                for (int col = 0; col < cols; col++)
                {
                    Console.Write(result[row, col] + " ");
                }
                Console.WriteLine();
            }
        }
    }
}

File: MyLearning/PythonPractice/dsa_practice/arrays/convert2d_to1d.py
"""2D/1D array conversions."""

# C# Convert2DTo1D.Convert2DArrayTo1D

def convert2d_array_to1d(matrix=None):
    """Flatten a 2D matrix into a 1D list in row-major order.

    Problem: Convert a 2D array into a 1D array.
    Approach: Traverse rows and columns and append to result.
    Time complexity: O(r * c)
    Space complexity: O(r * c)
    """
    if matrix is None:
        matrix = [
            [1, 1, 1, 1],
            [1, 0, 1, 1],
            [1, 1, 0, 1],
            [1, 0, 0, 1],
        ]

    result = []
    for row in matrix:
        for value in row:
            result.append(value)

    return result


# C# Convert2DTo1D.Convert1DArrayTo2D

def convert1d_array_to2d(flat_array=None, rows=4, cols=4):
    """Convert a 1D list into a 2D matrix in row-major order.

    Problem: Convert a 1D array into a 2D array with given dimensions.
    Approach: Fill rows and columns from the flat list.
    Time complexity: O(r * c)
    Space complexity: O(r * c)
    """
    if flat_array is None:
        flat_array = [
            1, 1, 1, 1,
            1, 0, 1, 1,
            1, 1, 0, 1,
            1, 0, 0, 1,
        ]

    if len(flat_array) < rows * cols:
        raise IndexError("flat_array is too short for the given dimensions")

    result = [[0 for _ in range(cols)] for _ in range(rows)]
    index = 0

    for row in range(rows):
        for col in range(cols):
            result[row][col] = flat_array[index]
            index += 1

    return result
